### ISSUE TITLE
encrypted api key password only work through TF_VAR_private_key_password environment variable, but not the pass_phrase setting in config file

### ISSUE THREAD

**Issue by luckeyca** [oracle/terraform-provider-oci#1663]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version and Provider Version
Terraform v1.2.8
on linux_amd64
+ provider registry.terraform.io/oracle/oci v4.86.1

### Affected Resource(s)

All. This is related to terraform oci authentication

### Terraform Configuration Files

```hcl
N/A
```

### Expected Behavior

When using encrypted API key to authenticate, terraform should be able to use either TF_VAR_private_key_password environment variable, OR the pass_phrase setting in ~/.oci/config file. which is how oci cli works.

oci cli works with either OCI_CLI_PASSPHRASE environment variable or pass_phrase setting in ~/.oci/config file.

### Actual Behavior

Only TF_VAR_private_key_password environment variable, but NOT pass_phrase setting in ~/.oci/config file

### Steps to Reproduce

1. define private_key_password = var.private_key_password in provider block.
2. in ~/.oci/config file, set pass_phrase for the encrypted api key, BUT NOT set the TF_VAR_private_key_password environment variable, when running terraform command, it will prompt for private key password.

**Comment by MeharwadeDivya** [comment]
Issue is not reproducible. Current provider behavior is working as expected. 
Steps tried:
- Start with a fresh env, type env and make sure no TF_VAR variables exist.
- export TF_LOG=TRACE to check logs
- create a password protected keypair
- ensure your oci config does not have any default config or your tf file have any parameter values added.
- export only the necessary TF_VARs
`TF_VAR_tenancy_ocid=ocid1.tenancy.oc1.xxxxxx
TF_VAR_fingerprint=11:12:13:14:15:16:17:18:19:20:21:22:23:24:25:26
TF_VAR_user_ocid=ocid1.user.oc1..xxxxxx
TF_VAR_region=us-phoenix-1
TF_VAR_private_key_password=1234`
- terraform init
- terraform plan - plan succeeded

`2023/05/09 19:23:20 [INFO] backend/local: plan operation completed

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # oci_core_vcn.internal will be created
  + resource "oci_core_vcn" "internal" {
      + byoipv6cidr_blocks               = (known after apply)
      + cidr_block                       = "172.16.0.0/20"
      + cidr_blocks                      = (known after apply)
2023/05/09 19:23:20 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info
      + compartment_id                   = "ocid1.tenancy.oc1..aaaaaaaawgbzspd6cjovpmic7dahwbalg4opcbixgguo6z2s7t335m6lalbq"
2023/05/09 19:23:20 [TRACE] statemgr.Filesystem: unlocking terraform.tfstate using fcntl flock
      + default_dhcp_options_id          = (known after apply)
      + default_route_table_id           = (known after apply)
      + default_security_list_id         = (known after apply)
      + defined_tags                     = (known after apply)
      + display_name                     = "My internal VCN"
      + dns_label                        = "internal"
      + freeform_tags                    = (known after apply)
      + id                               = (known after apply)
      + ipv6cidr_blocks                  = (known after apply)
      + ipv6private_cidr_blocks          = (known after apply)
      + is_ipv6enabled                   = (known after apply)
      + is_oracle_gua_allocation_enabled = (known after apply)
      + state                            = (known after apply)
      + time_created                     = (known after apply)
      + vcn_domain_name                  = (known after apply)

      + byoipv6cidr_details {
          + byoipv6range_id = (known after apply)
          + ipv6cidr_block  = (known after apply)
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.`

**Comment by luckeyca** [comment]
Hi @MeharwadeDivya you completely misunderstood/mis-read the problem I reported here. As I said, the problem is with pass_phrase setting in the ~/.oci/config file when TF_VAR_private_key_password environment variable is NOT set. The TF_VAR_private_key_password  itself works. You tested the wrong thing! Please read the problem description and follow the step-to-reproduce. @ravinitp  Please re-open this ticket as it's not resolved at all.

**Comment by ravinitp** [comment]
Hi,

I have attempted to reproduce the issue you reported in the following manner:

1. Created an encrypted API key pair.

2. Generated a configuration file with the default profile:

```
[DEFAULT]
user=ocid1.user.oc1..xxxxxxx
fingerprint=XX:XX:XX:XX
tenancy=ocid1.tenancy.oc1..Xxxxxxxxxxxxx
region=us-ashburn-1
key_file=<private key path>
pass_phrase=<password>
```
Executed terraform apply on the provided example and confirmed that it worked as expected.
Based on this, I would like to suggest the following to the customer:

It appears that the customer has set the required keys and OCIDs as environment variables within the bash shell. Therefore, I recommend unsetting all the environment variables with the prefixes TF_VAR_user_ocid), TF_VAR_tenancy_ocid, TF_VAR_fingerprint, TF_VAR_private_key etc. 
Please remove or unset these variables and then retry the operation.

I hope this helps! Let me know if you have any further questions.

**Comment by luckeyca** [comment]
Hi, in that case, while your workaround works, you are not following oracle documented authentication flow for terraform providers. As this documentation here: https://docs.oracle.com/en-us/iaas/Content/API/SDKDocs/terraformproviderconfiguration.htm, the authentication needs to follow the order of the precedence when authentication information are defined in multiple places. See the bottom of the documentation for details.

I filed a similar bug report on the exact same day for the same bug to OCI CLI team on github, they fixed in last Nov. 

Basically in the terraform provider authentication logic, it needs to check all the authentication methods following the order of precedence as defined in the documentation. Please have your developers looked into it.

The oci cli bug was here: https://github.com/oracle/oci-cli/issues/586
